File: src/main/java/pl/training/module7/breakmethod/before/LegacyReleaseManifestBuilder.java
package pl.training.module7.breakmethod.before;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;
import java.util.StringJoiner;

import pl.training.module7.breakmethod.ManifestEntry;

public final class LegacyReleaseManifestBuilder {
    public String build(List<ManifestEntry> entries) {
        Objects.requireNonNull(entries, "entries");

        var validatedEntries = new ArrayList<ManifestEntry>(entries.size());
        for (ManifestEntry entry : entries) {
            Objects.requireNonNull(entry, "entries must not contain null");
            Objects.requireNonNull(entry.artifact(), "artifact");
            if (entry.artifact().isBlank()) {
                throw new IllegalArgumentException("artifact must not be blank");
            }
            Objects.requireNonNull(entry.checksum(), "checksum");
            if (entry.checksum().isBlank()) {
                throw new IllegalArgumentException("checksum must not be blank");
            }
            if (entry.deploymentOrder() < 0) {
                throw new IllegalArgumentException(
                        "deploymentOrder must not be negative");
            }
            validatedEntries.add(entry);
        }

        validatedEntries.sort(
                Comparator.comparingInt(ManifestEntry::deploymentOrder)
                        .thenComparing(ManifestEntry::artifact));

        var manifest = new StringJoiner("\n");
        for (ManifestEntry entry : validatedEntries) {
            manifest.add(
                    entry.deploymentOrder()
                            + "|"
                            + entry.artifact()
                            + "|"
                            + entry.checksum());
        }
        return manifest.toString();
    }
}

File: typescript/src/module7/breakmethod/before/LegacyReleaseManifestBuilder.ts
import { IllegalArgumentError } from '../../../shared/errors.js';
import { requireNonNull } from '../../../shared/requireNonNull.js';
import type { ManifestEntry } from '../ManifestEntry.js';

export class LegacyReleaseManifestBuilder {
  build(entries: readonly ManifestEntry[]): string {
    requireNonNull(entries, 'entries');

    const validatedEntries: ManifestEntry[] = [];
    for (const entry of entries) {
      requireNonNull(entry, 'entries must not contain null');
      requireNonNull(entry.artifact, 'artifact');
      if (entry.artifact.trim() === '') {
        throw new IllegalArgumentError('artifact must not be blank');
      }
      requireNonNull(entry.checksum, 'checksum');
      if (entry.checksum.trim() === '') {
        throw new IllegalArgumentError('checksum must not be blank');
      }
      if (entry.deploymentOrder < 0) {
        throw new IllegalArgumentError(
          'deploymentOrder must not be negative');
      }
      validatedEntries.push(entry);
    }

    validatedEntries.sort((a, b) =>
      a.deploymentOrder - b.deploymentOrder
      || (a.artifact < b.artifact ? -1 : a.artifact > b.artifact ? 1 : 0));

    const manifest: string[] = [];
    for (const entry of validatedEntries) {
      manifest.push(
        entry.deploymentOrder
        + '|'
        + entry.artifact
        + '|'
        + entry.checksum);
    }
    return manifest.join('\n');
  }
}
